chamber/models/__init__.py

```diff
@@ -202,28 +203,75 @@ def _persistence_clean(self):
             else:
                 raise PersistenceException(', '.join(map(force_text, er.messages)))
 
-    def save(self, clean_before_save=None, clean_after_save=None, clean_before_delete=None, clean_after_delete=None,
-             force_insert=False, force_update=False, using=None, update_fields=None, *args, **kwargs):
-        clean_before_save = self._smart_meta.clean_before_save if clean_before_save is None else clean_before_save
+    def _get_save_extra_kwargs(self):
+        return {}
+
+    def _pre_save(self, *args, **kwargs):
+        pass
+
+    def _save(self, clean_pre_save=None, clean_post_save=None, force_insert=False, force_update=False, using=None,
+              update_fields=None, *args, **kwargs):
+        clean_pre_save = self._smart_meta.clean_pre_save if clean_pre_save is None else clean_pre_save
+        clean_post_save = self._smart_meta.clean_post_save if clean_post_save is None else clean_post_save
+
         change = bool(self.pk)
         changed_fields = set(self.changed_fields)
-        self.pre_save(change, changed_fields, *args, **kwargs)
-        if clean_before_save:
-            try:
-                self.full_clean()
-            except ValidationError as er:
-                if hasattr(er, 'error_dict'):
-                    raise PersistenceException(', '.join(
-                        ('%s: %s' % (key, ', '.join(map(force_text, val))) for key, val in er.message_dict.items())))
-                else:
-                    raise PersistenceException(', '.join(map(force_text, er.messages)))
+        kwargs.update(self._get_save_extra_kwargs())
+
+        self._pre_save(change, changed_fields, *args, **kwargs)
+
+        if clean_pre_save:
+            self._clean_pre_save()
+
         super(SmartModel, self).save(force_insert=force_insert, force_update=force_update, using=using,
                                      update_fields=update_fields)
-        self.post_save(change, changed_fields, *args, **kwargs)
 
-    def post_save(self, *args, **kwargs):
+        self._post_save(change, changed_fields, *args, **kwargs)
+
+        if clean_post_save:
+            self._clean_post_save()
+
+    def _post_save(self, *args, **kwargs):
+        pass
+
+    def save(self, *args, **kwargs):
+        if self._smart_meta.atomic_save:
+            with transaction.atomic():
+                self._save(*args, **kwargs)
+        else:
+            self._save(*args, **kwargs)
```

Comment:
Save takes care of the atomicity. Therefore, you cannot overload it if there is some behaviour you wish to add to save, it would be outside of the transaction...